### ISSUE TITLE
How to use the test function from the trainer?

### ISSUE THREAD

**Issue by LuposX** [PyTorchLightning/pytorch-lightning#1088]
#### What is your question?

So when I `test `my model with `trainer.test()` I get as output this:
```
 ----------------------------------------------------------------------------------------------------
TEST RESULTS
{}
----------------------------------------------------------------------------------------------------
```
I suppose in the empty bracket can stand sth. But I dont know what can stand there and where to do it. I can't seem to find it in the docs not under [trainer](https://pytorch-lightning.readthedocs.io/en/0.7.1/trainer.html#testing) nor under [test set](https://pytorch-lightning.readthedocs.io/en/0.7.1/test_set.html) stands for what this is. If I am right and it doesn'T stand anywhere maybe add to the docs.

#### Code

`trainer.test()`

#### What have you tried?
Searching the docs and google it.

#### What's your environment?

 - Windows 10
 - conda
 - lighning 0.7.1

**Comment by williamFalcon** [comment]
you have to return something in test_epoch_end. right now you return an empty dict

**Comment by LuposX** [comment]
@williamFalcon 
I did return sth in `test_end` than I switched the method out to `test_epoch_end`. Thats my code:
```
    def test_epoch_end(self, outputs):
        avg_acc = 100 * self.test_correct_counter / self.test_total_counter
        avg_loss = torch.stack([x['test_loss'] for x in outputs]).mean()

        self.test_correct_counter = 0
        self.test_total_counter = 0

        tensorboard_logs = {'avg_acc': avg_acc, 'test_loss': avg_loss}
        return {'avg_test_loss': avg_loss, 'avg_acc': avg_acc, 'log': tensorboard_logs}
```
But it still returns: 
 ```
----------------------------------------------------------------------------------------------------
TEST RESULTS
{}
----------------------------------------------------------------------------------------------------
```

**Comment by polars05** [comment]
@LuposX, I'm doing it as follows:
```
class CoolSystem(pl.LightningModule):

    def __init__(self, hparams):
        super(CoolSystem, self).__init__()

        ########## define the model ########## 
        ...
        ########## /define the model ########## 

    ...
    
    def test_step(self, batch, batch_idx):
        # OPTIONAL
        x, y = batch
        y_hat = self.forward(x)

        _, preds = torch.max(y_hat, 1)
        correct = torch.sum(preds == y.data)

        return {'test_loss': F.cross_entropy(y_hat, y), 'test_correct': correct}

    def test_end(self, outputs):
        # OPTIONAL
        avg_loss = torch.stack([x['test_loss'] for x in outputs]).mean()

        avg_acc = torch.stack([x['test_correct'].float() for x in outputs]).sum() / len(self.test_dataloader().dataset)

        logs = {'test_loss': avg_loss, 'test_acc': avg_acc}
        
        return {'avg_test_loss': avg_loss, 'avg_test_acc': avg_acc, 'log': logs, 'progress_bar': logs}

    ...

    @pl.data_loader
    def test_dataloader(self):
      transform = transforms.Compose([
                              transforms.Resize(256),
                              transforms.CenterCrop(224),
                              transforms.ToTensor(),
                              transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
                              ])
                            
      test_set = torchvision.datasets.ImageFolder(os.path.join(self.data_dir, 'val'), transform)
      test_loader = torch.utils.data.DataLoader(val_set, batch_size=8, shuffle=False, num_workers=4)

      return test_loader
```

And I ran the test by:
```
model_test = CoolSystem.load_from_checkpoint('/path/to/checkpoint/filename.ckpt') 

trainer = pl.Trainer()

trainer.test(model_test)
```

And I get:
```
----------------------------------------------------------------------------------------------------
TEST RESULTS
{'test_loss': tensor(0.4956), 'test_acc': tensor(0.9020)}
----------------------------------------------------------------------------------------------------
```

Hopefully it helps you!

**Comment by LuposX** [comment]
@polars05 
You shouldn't use `test_end()` instead you should use `test_epoch_end()` same with `test_step()`.
'test_end was deprecated in 0.7.0 and will be removed 1.0.0.  'Use test_epoch_end instead.'
'validation_end was deprecated in 0.7.0 and will be removed 1.0.0  'Use validation_epoch_end instead.'

**Comment by LuposX** [comment]
So you need to set `'progress_bar'` in the return statement of `test_epoch_end` to get a return value.

**Comment by isolet** [comment]
@LuposX I'm facing the same issue. Why inside 'progress_bar' instead of the return value itself?

**Comment by LuposX** [comment]
@isolet I dont know. Its just when I set it I get a return value and when not than not. :)

**Comment by tcwalther** [comment]
I would like to vote for reopening this issue, in particular because the Introduction Guide suggests that any value returned in the dictionary in `test_epoch_end` will be displayed in the summary. I assume that this was the case in the past, and that the Introduction Guide is simply outdated in this regard. Nevertheless, having to wrap it in a key called `progress_bar` seems counter-intuitive.

More generally speaking, it would help if there was a centralised documentation on what keys to return in the dictionary in each of these functions :-).

And of course, thanks so much for the work on this library. I'm just exploring it right now, but it looks really nice to use :-). I'm quite excited!

**Comment by kiansierra** [comment]
Hi, I was getting the same issue trying to get an output from `on_epoch_end`

I was hoping to do `results = trainer.test(model, datamodule) `

I finally settled with `def test_epoch_end(self, outputs):
        results = process_outputs(outputs)
        self.test_results = results
        return results
results = model.test_results`

**Comment by whoknowsB** [comment]
> Hi, I was getting the same issue trying to get an output from `test_epoch_end`
> 
> I was hoping to do
> 
> ```
> def test_epoch_end(self, outputs): 
>       results = process_outputs(outputs)
>       return results 
> 
> results = trainer.test(model, datamodule) 
> ```
> 
> I finally settled with
> 
> ```
> def test_epoch_end(self, outputs): 
>       results = process_outputs(outputs)
>       self.test_results = results
>       return results 
> 
> trainer.test(model, datamodule)
> results = model.test_results
> ```

Guys, any news on this? This still persists. I don't think this is the right way to go. Is it?

**Comment by jumpynitro** [comment]
I would like to vote for reopening this issue, the solution is too dark right now. I think pytorch lightning has a lot of potential. A better solution is needed

**Comment by bsridatta** [comment]
> test_epoch_end

Not working please reopen!

**Comment by mfoglio** [comment]
> I think you have to call `self.log_dict(results)` at the end of `test_epoch_end`. I don't see this documented explicitly anywhere, though.

It seems that calling ` self.log_dict(log_metrics)` even inside `test_step` does the job.

**Comment by FrankwaP** [comment]
It seems one needs to use `predict_step` (and `trainer.predict`) in order to explicitly return something (without using "tricks" like logging).

For testing the output is calculated in `evaluation_epoch_loop.EvaluationLoop.on_run_end` method which:
1. does not expect any results from `test_epoch_end` (named `_evaluation_epoch_end` here)
2. only cares about logged outputs (`_logged_outputs`) not the returned outputs (`_outputs`)

```python
    def on_run_end(self) -> List[_OUT_DICT]:
        """Runs the ``_on_evaluation_epoch_end`` hook."""
        # if `done` returned True before any iterations were done, this won't have been called in `on_advance_end`
        self.trainer._logger_connector.epoch_end_reached()

        # hook
        self._evaluation_epoch_end(self._outputs)   # 1. no result expected from  test_epoch_end
        self._outputs = []  # free memory  # 2. does not care about the outputs of test_step

        # hook
        self._on_evaluation_epoch_end()

        logged_outputs, self._logged_outputs = self._logged_outputs, []  # 2. only cares about logged outputs
        # include any logged outputs on epoch_end
        epoch_end_logged_outputs = self.trainer._logger_connector.update_eval_epoch_metrics()
        all_logged_outputs = dict(ChainMap(*logged_outputs))  # list[dict] -> dict
        all_logged_outputs.update(epoch_end_logged_outputs)
        for dl_outputs in logged_outputs:
            dl_outputs.update(epoch_end_logged_outputs)

        # log metrics
        self.trainer._logger_connector.log_eval_end_metrics(all_logged_outputs)

        # hook
        self._on_evaluation_end()

        # enable train mode again
        self._on_evaluation_model_train()

        if self.verbose and self.trainer.is_global_zero:
            assert self.trainer.state.stage is not None
            self._print_results(logged_outputs, self.trainer.state.stage)

        return logged_outputs
```

It's been coded with the sole purpose of displaying the logged information like this

> 
> ────────────────────────────────────────────────────────────────────────────────
>        Test metric             DataLoader 0
> ────────────────────────────────────────────────────────────────────────────────
>     test_metric_suprv       0.1450386643409729
> ────────────────────────────────────────────────────────────────────────────────


While the same method is very simple for `predict`:
```python
    def on_run_end(self) -> Optional[_PREDICT_OUTPUT]:
        """Calls ``on_predict_epoch_end`` and ``on_predict_end`` hooks and returns results from all dataloaders."""
        results = self._on_predict_epoch_end()
        self._on_predict_end()
        return results
```

**Comment by Abermal** [comment]
The issue is still there.

**Comment by FrankwaP** [comment]
> Right there, there is no way to save predictions from the model which does not feel hacky.

What's wrong with simply using `predict` ?

**Comment by kasnerz** [comment]
> > Right there, there is no way to save predictions from the model which does not feel hacky.
> 
> What's wrong with simply using `predict` ?

You are right, I actually discovered this solution now. I guess this function did not exist in [early versions](https://pytorch-lightning.readthedocs.io/en/1.2.5/common/trainer.html) of PyTorch Lightning and later I did not notice it was added.

**Comment by MannixWong** [comment]
> > Right there, there is no way to save predictions from the model which does not feel hacky.
> 
> What's wrong with simply using `predict` ?

Are these two methods the same? I saw someone say "test" for the data with labels in test set, and "predict" for the ever seen data without labels.

**Comment by FrankwaP** [comment]
No they're not the same, look at my previous comment (oct 3). Yes, `test` needs labelled data. But you do whatever you want with `predict`, you can even use it with your train data if you want to visually check the predictions.

**Comment by adosar** [comment]
> I think you have to call `self.log_dict(results)` at the end of `test_epoch_end`. I don't see this documented explicitly anywhere, though.

This is the weakest point of `pytorch-lightning` IMO. So many functionalities, but if we need to do "black magic" instead of looking at the official tutorials and the API, then something is wrong. Many essential parts lack thorough and explicit documentation.
